source/common/http/conn_pool_base.cc

```diff
@@ -1,40 +1,293 @@
-  // We don't track any failure state, as the client should be deferred destroyed once a failure
-  // event is handled.
-  if (connect_timer_) {
-    return Connecting;
+void ConnPoolImplBase::checkForDrained() {
+  if (drained_callbacks_.empty()) {
+    return;
   }
 
-  return Connected;
+  closeIdleConnections();
+
+  if (pending_requests_.empty() && ready_clients_.empty() && busy_clients_.empty()) {
+    ENVOY_LOG(debug, "invoking drained callbacks");
+    for (const DrainedCb& cb : drained_callbacks_) {
+      cb();
+    }
+  }
+}
+
+void ConnPoolImplBase::onConnectionEvent(ConnPoolImplBase::ActiveClient& client,
+                                         Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    // The client died.
+    ENVOY_CONN_LOG(debug, "client disconnected, failure reason: {}", *client.codec_client_,
+                   client.codec_client_->connectionFailureReason());
+
+    Envoy::Upstream::reportUpstreamCxDestroy(host_, event);
+    const bool incomplete_request = client.closingWithIncompleteRequest();
+    if (incomplete_request) {
+      Envoy::Upstream::reportUpstreamCxDestroyActiveRequest(host_, event);
+    }
+
+    if (client.state_ == ActiveClient::State::CONNECTING) {
+      host_->cluster().stats().upstream_cx_connect_fail_.inc();
+      host_->stats().cx_connect_fail_.inc();
+
+      // Raw connect failures should never happen under normal circumstances. If we have an upstream
+      // that is behaving badly, requests can get stuck here in the pending state. If we see a
+      // connect failure, we purge all pending requests so that calling code can determine what to
+      // do with the request.
+      // NOTE: We move the existing pending requests to a temporary list. This is done so that
+      //       if retry logic submits a new request to the pool, we don't fail it inline.
+      purgePendingRequests(client.real_host_description_,
+                           client.codec_client_->connectionFailureReason());
+    }
+
+    // We need to release our resourceManager() resources before checking below for
+    // whether we can create a new connection. Normally this would happen when
+    // client's destructor runs, but this object needs to be deferredDelete'd(), so
+    // this forces part of its cleanup to happen now.
+    client.releaseResources();
+
+    dispatcher_.deferredDelete(client.removeFromList(owningList(client.state_)));
+    if (incomplete_request) {
+      checkForDrained();
+    }
+
+    client.state_ = ActiveClient::State::CLOSED;
+
+    // If we have pending requests and we just lost a connection we should make a new one.
+    if (!pending_requests_.empty() &&
+        host_->cluster().resourceManager(priority_).requests().canCreate() &&
+        host_->cluster().resourceManager(priority_).connections().canCreate()) {
+      createNewConnection();
+    }
```

Comment:
Coming back to this discussion, if the circuit breaker check here fails, what happens? I think we could still starve? (This is the block that used to create a connection no matter what, right? Or is that somewhere else)?

Reply:
No, the block that unconditionally makes the connection didn't move; it was always in `newStream()`. In the original http1 pool code, it has this check for this case: `if (pending_requests_.size() > (ready_clients_.size() + busy_clients_.size())) {`. I think the presumption is that if we just closed a connection, there should be available circuit breaker capacity for 1 more connection.

In the new version, I think it would only be starved if there are no connections in this pool, so I think it needs roughly the same check as in `newStream()`. I'll sort that out.

Reply:
Also, while thinking this through, I don't think the requests circuit breaker should be checked here at all. The behavior has always been that if a connection is ready, then requests are attached to it, and at that time if the requests circuit breaker is exceeded, an `onPoolFailure(Overflow)` is issued. That code happens in `attachRequestToClient()`, so I think that's the only place we should check that circuit breaker.